### ISSUE TITLE
Serve command loads server in wrong environment

### ISSUE THREAD

**Issue by dinandmentink** [laravel/framework#27828]
- Laravel Version: 5.8.3
- PHP Version: 7.2.15
- Database Driver & Version: sqlite locally (irrelevant I believe)

### Description:
When using the artisan serve command the correct environment is not loaded.

### Steps To Reproduce:

1. run `php artisan serve --env=testing` with a `.env.testing` file with `APP_ENV=testing`
2. put a `{{ env("APP_ENV") }}` in any view or anything
3. Instead of `testing`, see `local`

It seems the environment *is* correctly loaded for artisan commands, but not for the resulting server process. Contrast the above with:

1. run `php artisan tinker --env=testing` with a `.env.testing` file with `APP_ENV=testing`
2. run `env("APP_ENV")`
3. result is as one would expect: `testing`

**Comment by mllnd** [comment]
Did you try resetting the configuration cache: `php artisan config:clear`

**Comment by driesvints** [comment]
This works for me. Can you first please try one of the following support channels? If you can actually identify this as a bug, feel free to report back.

- [Laracasts Forums](https://laracasts.com/discuss)
- [Laravel.io Forums](https://laravel.io/forum)
- [StackOverflow](https://stackoverflow.com)
- [Discord](https://discordapp.com/invite/KxwQuKb)
- [Larachat](https://larachat.co)
- [IRC](https://webchat.freenode.net/?nick=laravelnewbie&channels=%23laravel&prompt=1)

**Comment by dinandmentink** [comment]
> Try resetting the configuration cache: `php artisan config:clear`

Thanks. I have done this. 

I do believe this to be a bug. Steps to reproduce based solely on `laravel/laravel`:

1. `composer create-project --prefer-dist laravel/laravel test`
2. copy `.env` to `.env.testing`
3. change in `.env.testing` APP_ENV to `testing`
4. add `<h1> {{ env('APP_ENV') }} </h1>` to welcome.blade.php
5. run `php artisan serve --env=testing`

The page will show `local`, not the expected `testing`. From within tinker however the correct `testing` will be shown.

If still unconvinced this is a bug I'll go to one of the support channels.

**Comment by staudenmeir** [comment]
I can reproduce this.

**Comment by dinandmentink** [comment]
> I can reproduce this. It works in Laravel 5.7.

Correct. I ran into this in 3 apps upon upgrade to 5.8(.3).

**Comment by driesvints** [comment]
That's very odd. I just tested again and I can indeed reproduce it. Not sure why it didn't the first time I tried.

**Comment by driesvints** [comment]
I really can't see what changed it. Nothing really has changed to the `ServeCommand`.

**Comment by staudenmeir** [comment]
Maybe this is caused by #27462 and the related PRs. /cc @GrahamCampbell

**Comment by rs-sliske** [comment]
isnt the env helper meant to only be used in config files?

(not sure if actually related or not)

**Comment by staudenmeir** [comment]
@rs-sliske We are only using it to demonstrate the issue. `config('app.env')` doesn't work either.

**Comment by dinandmentink** [comment]
> @rs-sliske We are only using it to demonstrate the issue. `config('app.env')` doesn't work either.

Correct. Problem is not in the env helper as far as I can see. If I drop a psyshell in `index.php` it seems that after constructing the `$app` the actual `environmentFile()` and related are off. Something in booting the app seems to be wrong, causing the env to be set to the .env instead of the correct file.

To unfamiliar with the inner workings of laravel to help much I'm afraid.

**Comment by staudenmeir** [comment]
This is caused by the different ways Laravel 5.7 and 5.8 set environment variables.

- [5.7](https://github.com/vlucas/phpdotenv/blob/2.6/src/Loader.php#L339-L344) sets the variables in `$_ENV` and `$_SERVER` *and* calls `putenv()`.
- [5.8](https://github.com/laravel/framework/blob/ed22856cc26524ef4cc6db1d34d9f1061c34f998/src/Illuminate/Foundation/Bootstrap/LoadEnvironmentVariables.php#L91) only sets the variables in `$_ENV` and `$_SERVER`.

Adding the `putenv()` call to Laravel 5.8 fixes the issue: https://github.com/staudenmeir/framework/commit/1412636087d517842991c1817b230c3faf83b9cd

All the tests still work. Can you approve this @GrahamCampbell?

**Comment by staudenmeir** [comment]
**UPDATE**

Apparantely, it's not that simple: Now removing `E` from the `variables_order` setting in `php.ini` completely breaks environment handling.

**Comment by xendk** [comment]
I tried upgrading Lumen, and it broke my app for me. The problem was that it wasn't seeing a required env variable set with `export` before starting the app.

I gather that Dotenv switched from using `getenv` to `$_ENV` due to threading issues with the former, the problem is that `$_ENV` only gets populated with the existing environment if `E` is part of `variables_order` which isn't standard. Standard PHP install on Ubuntu doesn't include it (I believe it's a security measure). A bit of asking around shows that  and at least some OSX installs doesn't either, and for docker containers it's a tossup (pretty much everything based on Ubuntu will probably have the problem).

I think that not using `put`/`getenv` was an explicit choice, but considering the amount of people that'll get hit by this, whether some compatibility fix can be done. Maybe preloading `$_ENV` from `getenv` if the former is empty when starting up. @GrahamCampbell is probably the person to talk to.

**Comment by driesvints** [comment]
@staudenmeir `PutenvAdapter` was specifically removed to make the `env` helper read-only. Can we fix this any other way?

**Comment by GrahamCampbell** [comment]
Why is it adding putenv fixes this\/

**Comment by xendk** [comment]
@GrahamCampbell it's not putenv that fixes it, but the fact that PutenvAdapter uses getenv to read env variables.

**Comment by staudenmeir** [comment]
@xendk You need both `getenv()` and `putenv()` to fix the issue.

**Comment by staudenmeir** [comment]
@driesvints Why is `getenv()` not read-only?

@GrahamCampbell I don't fully understand it. Somehow, `php -S` treats `getenv()` and `$_ENV`/`$_SERVER` differently.

**Comment by xendk** [comment]
@staudenmeir You do? Why? The problem with PutenvAdapter is that it shadows EnvConstAdapter for setting too so all variables is then set with `putenv()` and doesn't show up in `$_ENV` anymore.

**Comment by staudenmeir** [comment]
@xendk `php artisan serve --env` only works for me when I add the `PutenvAdapter` to `LoadEnvironmentVariables::createDotenv()` and `env()`.

**Comment by xendk** [comment]
@StanAngeloff yeah, and that completely shadows `EnvConstAdapter`. You might want to try and hack `EnvConstAdapter::get` to try using `getenv` if the key was not found in `$_ENV` instead, and see if that doesn't fix things too.

**Comment by nomad-software** [comment]
I'm suffering the same issue here:
https://github.com/laravel/framework/issues/27949

Removing the `PutenvAdapter ` adapter breaks loads of third-party composer libraries too.

**Comment by nomad-software** [comment]
Here's another issue related to this:

https://github.com/laravel/framework/issues/27913

**Comment by staudenmeir** [comment]
This has been fixed in the latest release 5.8.7.

**Comment by dinandmentink** [comment]
> This has been fixed in the latest release 5.8.7.

Thanks a lot!

**Comment by driesvints** [comment]
Thanks for everyone's feedback and help with this.

**Comment by hopeseekr** [comment]
I have liberated Laravel 5.7's `env()` from the dustbin of the git repo and made it its own package. What's more, using some real composer hacks, I have ensured that it will always be autoloaded first, overriding whatever Laravel's current env() does. 

Benefits:
* Works with *any* package, even non-Laravel ones (it was my secondary reason)
* You no longer have to fear Laravel just going one day "Oops! We're not going to read from the environment anymore! Too bad!" on a non-major release.

https://github.com/phpexpertsinc/Laravel57-env-polyfill

To install: `composer require phpexperts/laravel-env-polyfill`.

**Comment by martijnhiemstra** [comment]
It is now Augustus 2022 and this is still a issue! This hasnt't been fixed yet!! Does anybody have a solution?

**Comment by DatasoftSolutionsNet** [comment]
I'm having an issue with this as well.

**Comment by hopeseekr** [comment]
I created a solution for this years and years ago, back in 2019. 

I ported Laravel 5.7's env() to be usable by any app. It even overrides Laravel's implementation, if the package is included:

    composer require phpexperts/env

Then just `env('MY_VAR');` like you would in pre-5.8...
